File: app/src/main/java/io/neft/Renderer/NativeItem.java
package io.neft.Renderer;

import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.RectF;
import android.os.SystemClock;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;

import java.util.ArrayList;
import java.util.HashMap;

import io.neft.Client.Action;
import io.neft.Client.InAction;
import io.neft.Client.OutAction;
import io.neft.Client.Reader;
import io.neft.MainActivity;

public class NativeItem extends Item {
    abstract public static class NativeType {
        abstract public Item work(MainActivity app);
    }

    public static HashMap<String, NativeType> types = new HashMap<>();

    static void register(final MainActivity app) {
        app.client.actions.put(InAction.CREATE_NATIVE_ITEM, new Action() {
            @Override
            public void work(Reader reader) {
                final String ctor = reader.getString();
                final NativeType type = types.get(ctor);
                if (type == null) {
                    Log.w("Neft", "Native Item '" + ctor + "' type not found");
                    new NativeItem(app);
                } else {
                    type.work(app);
                }
            }
        });

        app.client.actions.put(InAction.ON_NATIVE_ITEM_POINTER_PRESS, new Action() {
            @Override
            public void work(Reader reader) {
                ((NativeItem) app.renderer.getItemFromReader(reader)).onPointerPress(reader.getFloat(), reader.getFloat());
            }
        });

        app.client.actions.put(InAction.ON_NATIVE_ITEM_POINTER_RELEASE, new Action() {
            @Override
            public void work(Reader reader) {
                ((NativeItem) app.renderer.getItemFromReader(reader)).onPointerRelease(reader.getFloat(), reader.getFloat());
            }
        });

        app.client.actions.put(InAction.ON_NATIVE_ITEM_POINTER_MOVE, new Action() {
            @Override
            public void work(Reader reader) {
                ((NativeItem) app.renderer.getItemFromReader(reader)).onPointerMove(reader.getFloat(), reader.getFloat());
            }
        });
    }

    public boolean autoWidth = true;
    public boolean autoHeight = true;
    protected View view;
    protected long invalidateDuration = 0;
    protected long onPointerPressInvalidateDuration = 0;
    protected long onPointerReleaseInvalidateDuration = 0;
    protected long onPointerMoveInvalidateDuration = 0;

    public NativeItem(MainActivity app) {
        super(app);
    }

    @Override
    public void setWidth(float val) {
        super.setWidth(val);
        this.autoWidth = val == 0;
        this.updateSize();
    }

    @Override
    public void setHeight(float val) {
        super.setHeight(val);
        this.autoHeight = val == 0;
        this.updateSize();
    }

    protected void updateSize() {
        if (view != null) {
            view.measure(app.view.getWidth(), app.view.getHeight());
            int width = autoWidth ? view.getMeasuredWidth() : Math.round(this.width);
            int height = autoHeight ? view.getMeasuredHeight() : Math.round(this.height);
            view.layout(0, 0, width, height);
            pushWidth(width);
            pushHeight(height);
        }
    }

    protected void pushWidth(float val) {
        if (autoWidth && this.width != val) {
            super.setWidth(val);
            app.client.pushAction(OutAction.NATIVE_ITEM_WIDTH);
            app.client.pushInteger(id);
            app.client.pushFloat(app.renderer.pxToDp(val));
        }
    }

    protected void pushHeight(float val) {
        if (autoHeight && this.height != val) {
            super.setHeight(val);
            app.client.pushAction(OutAction.NATIVE_ITEM_HEIGHT);
            app.client.pushInteger(id);
            app.client.pushFloat(app.renderer.pxToDp(val));
        }
    }

    protected void callTouchEventOnView(int type, float x, float y) {
        if (view != null) {
            MotionEvent event = MotionEvent.obtain(
                    SystemClock.uptimeMillis(),
                    SystemClock.uptimeMillis(),
                    type,
                    x,
                    y,
                    0
            );
            view.onTouchEvent(event);
        }
    }

    public void onPointerPress(float x, float y) {
        callTouchEventOnView(MotionEvent.ACTION_DOWN, x, y);
        if (onPointerPressInvalidateDuration > 0) {
            this.invalidateDuration = onPointerPressInvalidateDuration;
            invalidate();
        }
    }

    public void onPointerRelease(float x, float y) {
        callTouchEventOnView(MotionEvent.ACTION_UP, x, y);
        if (onPointerReleaseInvalidateDuration > 0) {
            this.invalidateDuration = onPointerReleaseInvalidateDuration;
            invalidate();
        }
    }

    public void onPointerMove(float x, float y) {
        callTouchEventOnView(MotionEvent.ACTION_MOVE, x, y);
        if (onPointerMoveInvalidateDuration > 0) {
            this.invalidateDuration = onPointerMoveInvalidateDuration;
            invalidate();
        }
    }

    @Override
    protected void measure(final Matrix globalMatrix, RectF viewRect, final ArrayList<RectF> dirtyRects, final boolean forceUpdateBounds) {
        super.measure(globalMatrix, viewRect, dirtyRects, forceUpdateBounds);
        if (invalidateDuration > 0) {
            this.invalidateDuration -= app.view.frameDelay;
            invalidate();
        }
    }

    @Override
    protected void drawShape(final Canvas canvas, final int alpha) {
        if (view != null) {
            view.draw(canvas);
        }
    }
}

File: runtimes/ios/Neft/NativeItem.swift
import UIKit

class NativeItem: Item {
    class NativeUIView: UIView {}

    static var types: Dictionary<String, () -> NativeItem> = [:]
    class var name: String { return "Unknown" }

    override class func register(){
        onAction(.createNativeItem) {
            (reader: Reader) in
            let ctorName = reader.getString()
            let ctor = types[ctorName]
            if ctor == nil {
                print("Native item '\(ctorName)' type not found")
                save(item: NativeItem())
            } else {
                save(item: ctor!())
            }
        }

        onAction(.onNativeItemPointerPress) {
            (item: NativeItem, reader: Reader) in
            item.onPointerPress(reader.getFloat(), reader.getFloat())
        }

        onAction(.onNativeItemPointerMove) {
            (item: NativeItem, reader: Reader) in
            item.onPointerMove(reader.getFloat(), reader.getFloat())
        }

        onAction(.onNativeItemPointerRelease) {
            (item: NativeItem, reader: Reader) in
            item.onPointerRelease(reader.getFloat(), reader.getFloat())
        }
    }

    private static func on<T: NativeItem>(
        _ type: String,
        _ name: String,
        _ handler: @escaping (T, [Any?]) -> Void
        ) {
        let funcName = "renderer\(type)\(self.name.uppercaseFirst)\(name.uppercaseFirst)"
        App.getApp().client.addCustomFunction(funcName) {
            (inputArgs: [Any?]) in
            var args = inputArgs
            let index = Int(round(args[0] as! CGFloat))
            let item = App.getApp().renderer.items[index] as! T
            args.removeFirst()
            handler(item, args)
        }
    }

    internal static func onSet<T: NativeItem>(
        _ propertyName: String,
        _ handler: @escaping (T, [Any?]) -> Void
        ) {
        on("Set", propertyName, handler)
    }

    internal static func onSet<T: NativeItem>(
        _ propertyName: String,
        _ handler: @escaping (T, Bool) -> Void
        ) {
        onSet(propertyName) {
            (item: T, args: [Any?]) in
            handler(item, args[0] as! Bool)
        }
    }

    internal static func onSet<T: NativeItem>(
        _ propertyName: String,
        _ handler: @escaping (T, CGFloat) -> Void
        ) {
        onSet(propertyName) {
            (item: T, args: [Any?]) in
            handler(item, args[0] as! CGFloat)
        }
    }

    internal static func onSet<T: NativeItem>(
        _ propertyName: String,
        _ handler: @escaping (T, Int) -> Void
        ) {
        onSet(propertyName) {
            (item: T, args: [Any?]) in
            handler(item, Int(args[0] as! CGFloat))
        }
    }

    internal static func onSet<T: NativeItem>(
        _ propertyName: String,
        _ handler: @escaping (T, UIColor?) -> Void
        ) {
        onSet(propertyName) {
            (item: T, val: Int?) in
            handler(item, val != nil ? Color.hexColorToUIColor(val!) : nil)
        }
    }

    internal static func onSet<T: NativeItem>(
        _ propertyName: String,
        _ handler: @escaping (T, Item?) -> Void
        ) {
        onSet(propertyName) {
            (item: T, val: Int?) in
            handler(item, val != nil && val! >= 0 ? renderer.items[val!] : nil)
        }
    }

    internal static func onSet<T: NativeItem>(
        _ propertyName: String,
        _ handler: @escaping (T, String) -> Void
        ) {
        onSet(propertyName) {
            (item: T, args: [Any?]) in
            handler(item, args[0] as! String)
        }
    }

    internal static func onCall<T: NativeItem>(
        _ funcName: String,
        _ handler: @escaping (T, [Any?]) -> Void
        ) {
        on("Call", funcName, handler)
    }

    internal static func onCreate<T: NativeItem>(
        handler: @escaping () -> T
        ) {
        let typeName = self.name
        assert(NativeItem.types[typeName] == nil, "NativeItem \(typeName) onCreate already called")
        NativeItem.types[typeName] = handler
    }

    var autoWidth = true
    var autoHeight = true
    override var keysFocus: Bool {
        didSet {
            if keysFocus {
                view.becomeFirstResponder()
            } else {
                view.resignFirstResponder()
            }
        }
    }

    override var width: CGFloat {
        didSet {
            autoWidth = width == 0
            updateSize()
        }
    }

    override var height: CGFloat {
        didSet {
            autoHeight = height == 0
            updateSize()
        }
    }

    override init(view: UIView = UIView()) {
        super.init(view: view)
    }

    override func didSave() {
        super.didSave()
        updateSize()
    }

    internal func updateSize() {
        let width = autoWidth ? view.frame.width : self.width
        let height = autoHeight ? view.frame.height : self.height
        pushWidth(width)
        pushHeight(height)
    }

    private func pushWidth(_ val: CGFloat) {
        if autoWidth && width != val {
            self.width = val
            pushAction(.nativeItemWidth, val)
        }
    }

    private func pushHeight(_ val: CGFloat) {
        if autoHeight && height != val {
            self.height = val
            pushAction(.nativeItemHeight, val)
        }
    }

    internal func onPointerPress(_ x: CGFloat, _ y: CGFloat) {
    }

    internal func onPointerRelease(_ x: CGFloat, _ y: CGFloat) {
    }

    internal func onPointerMove(_ x: CGFloat, _ y: CGFloat) {
    }

    internal func pushEvent(event: String, args: [Any?]?) {
        let eventName = "rendererOn\(type(of: self).name.uppercaseFirst)\(event.uppercaseFirst)"
        var clientArgs: [Any?] = [CGFloat(id)]
        if args != nil {
            clientArgs.append(contentsOf: args!)
        }
        App.getApp().client.pushEvent(eventName, args: clientArgs)
    }
}
